### PR TITLE
PR: Remove trailing spaces and add script/module to automate the process

### PR DESCRIPTION
**Pull request by CAM-Gerlach**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)

In two separate commits, as requested, this PR removes trailing whitespace on both blank, and non-blank lines, from files with extensions on a manually curated whitelist, and with exclusions for any non-code test files and the vendored mathjax library. Also, the changes are checked against every open PR from the past >6 months, and any files that have changes that result in merge conflicts are excluded. They comprise the following:

<details>

```
PR #9226 has conflicts {'spyder/plugins/onlinehelp/onlinehelp.py', 'spyder/plugins/profiler/plugin.py', 'spyder/plugins/workingdirectory/plugin.py', 'spyder/plugins/findinfiles/plugin.py', 'spyder/plugins/breakpoints/plugin.py', 'spyder/plugins/console/plugin.py', 'spyder/plugins/pylint/plugin.py', 'spyder/plugins/outlineexplorer/plugin.py', 'spyder/plugins/variableexplorer/plugin.py', 'spyder/plugins/history/plugin.py', 'spyder/plugins/explorer/plugin.py'}
PR #9219 has conflicts {'spyder/plugins/outlineexplorer/widgets.py'}
PR #9127 has conflicts {'spyder/utils/workers/types.py'}
PR #9082 has conflicts {'spyder/plugins/outlineexplorer/widgets.py', 'spyder/utils/syntaxhighlighters.py'}
PR #9014 has conflicts {'spyder/plugins/outlineexplorer/widgets.py', 'spyder/utils/syntaxhighlighters.py'}
PR #8852 has conflicts {'spyder/plugins/variableexplorer/widgets/namespacebrowser.py'}
Total PR conflicts: 15 in 6
```

</details>

Finally, this work adds a comprehensive [script and module](https://github.com/spyder-ide/spyder/pull/9307/files#diff-71288fc955fd5bb4ed90b1943621795b), which can be accessed via APIs for both the command line and from other Python programs, that does the checks. It can be configured to either actually execute the changes, or merely warn about them and fail a build accordingly; and it can also optionally either scan the entire index, or various configurable subsets of modified files for use as a manual ``make check`` style script or an automated pre-commit hook. Many common components and infrastructure can likely be re-used in the other checker/fixer scripts, and it has been quite considerably tested over various scenarios (though by no means exhaustively).

Ideally, if #9226  could get merged quickly, that would take care of over 2/3rds of the remaining merge conflicts and I could re-base this one accordingly. Then, only 5 files remain, which could be handled on a PR by PR basis. #8734 would also be useful to get merged soon to help keep things clean until the final CI checks, pre-commit hooks and/or cleanup script can get implemented.

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: CAM-Gerlach

### CODE REVIEW

**Comment by ccordoba12** [comment]
@CAM-Gerlach, thanks a lot for your work on this!! I'm leaving this beta3 because we're trying to have everything ready to release beta2 next week.

We'll come back to review this afterwards (there are some whitespace removals that are unnecessary and the script should't be placed in `scripts`).

**Comment by CAM-Gerlach** [comment]
> I'm leaving this beta3 because we're trying to have everything ready to release beta2 next week.

Okay sure. If conflicts develop, I can always just regenerate it with the script.

> there are some whitespace removals that are unnecessary

Aren't *all* of them technically unnecessary)? I thought the point was to be consistent with removing them everywhere, so we can keep the "remove trailing whitespace on save" feature on and run ``yapf``, this script, etc to keep things clean without worrying about causing future conflicts.

> the script should't be placed in scripts

That's just where @goanpeca suggested I put it. Where should it go instead?

**Comment by ccordoba12** [comment]
> Aren't all of them technically unnecessary?

No, I already noted a few that are necessary, so I need to review this PR with care.

**Comment by ccordoba12** [comment]
> That's just where @goanpeca suggested I put it. Where should it go instead?

Ok, but the scripts in `scripts` are used to launch Spyder on Linux.

So it perhaps should go to `tools` in the root of the repo. And you can also move there `png-convert-and-optimize.sh`, which has been in the wrong place for a long time, but I didn't know where to put it.

**Comment by CAM-Gerlach** [comment]
> No, I already noted a few that are necessary

> there are some whitespace removals that are unnecessary

... :confounded: 

> Ok, but the scripts in `scripts` are used to launch Spyder on Linux.

Yeah, it did seem a little odd to put it there. There are also some random Windows scripts too.

> And you can also move there `png-convert-and-optimize.sh

Yes, that makes sense to put it wherever this goes. It definitely doesn't belong in the repo root like you said. We can put the other scripts and stuff in there so we can have a script to call everything at once. We could even integrate ``png_convert_and_optimize`` into the that architecture and automatically call it for any new or changed PNGs, along with our other stuff. Then link it all to a git pre-commit hook and never have to worry about it again...

**Comment by ccordoba12** [comment]
> Then link it all to a git pre-commit hook and never have to worry about it again...

Yeah, that'd be nice too.

It's your choice then for the new directory name. My suggestion are `tools` or `tooling`. Scikit-learn has `maint_tools`, which is good too.

**Comment by CAM-Gerlach** [comment]
> Yeah, that'd be nice too.

Out of scope here, ofc :) But that's the next step after we have the scripts working and everything harmonized.

> It's your choice then for the new directory name

I probably don't have any better suggestions than you, lol. @goanpeca , any thoughts?

**Comment by goanpeca** [comment]
Yes my bad, in my head and other repos the `scripts` folder is used for these. So hmmmmm I would go with `tools` for now and maybe at a `README.md` as well ? 

>So it perhaps should go to tools in the root of the repo. And you can also move there png-convert-and-optimize.sh, which has been in the wrong place for a long time, but I didn't know where to put it.

👍 

Thanks for working on this :-)

**Comment by goanpeca** [comment]
I will get back to review in more detail after beta2 is released! This is looking great 🎉

**Comment by goanpeca** [comment]
@CAM-Gerlach could you update this?

@ccordoba12 could you point to some of the spaces that were not supposed to be removed to check what the issue might be?

**Comment by CAM-Gerlach** [comment]
The only one I immediately see are the deliberate spaces in the issue template. Other than that, I'm not sure exactly what he meant, as I was rather conservative with this (not applying it to any non-actual-test files in the test directory, for instance, or any file types not on an explicit whitelist). However, its easy to add any files/paths that should not be converted to the blacklist in the script.

I can re-run it, but I'd rather wait to know what exclusions to add, since with the various Github calls, test-merging every PR and the substantial increase in unmerged PRs lately (with a number about to be merged), the runtime and resource cost isn't completely trivial and it would need to be redone anyway.

File: .github/ISSUE_TEMPLATE.md [hunk 1]
```diff
@@ -20,9 +20,9 @@
 
 ### What steps reproduce the problem?
 
-1. 
-2. 
-3.
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply .github/ISSUE_TEMPLATE.md]: As I mentioned above (specifically about this file), the way this is done is by adding the file to the blacklist in the script; its on a per-file basis. As such, I will add this file to the blacklist.

File: .github/ISSUE_TEMPLATE.md [hunk 2]
```diff
@@ -42,11 +42,11 @@ PASTE TRACEBACK HERE
 or (if Spyder won't launch) the "conda list" command
 from the Anaconda Prompt/Terminal/command line. --->
 
-* Spyder version: 
-* Python version: 
-* Qt version: 
-* PyQt version: 
-* Operating System name/version:
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply .github/ISSUE_TEMPLATE.md]: See above.

File: scripts/rm_whitespace.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply scripts/rm_whitespace.py]: Will do.

File: spyder/defaults/defaults-2.4.0.ini [hunk 4]
```diff
@@ -24,7 +24,7 @@ vertical_tabs = False
 use_custom_margin = True
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/defaults/defaults-2.4.0.ini]: > Revert all changes to this file 

See above. I'll add ``spyder/defaults/.*.ini``, to ignore all ``ini`` files in the defaults directly.

> It makes no sense to check for blanks in those files.

Why not?

That would miss other non-Spyder ``ini`` files, e.g. ``pytest.ini`` in the repo root. If someone injects spurious trailing spaces into that file, they would not get removed as they should by this script.

File: spyder/defaults/defaults-3.0.0.ini [hunk 5]
```diff
@@ -24,7 +24,7 @@ vertical_tabs = False
 use_custom_margin = True
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/defaults/defaults-3.0.0.ini]: See above.

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 6]
```diff
@@ -209,8 +209,8 @@ def long_banner(self):
             banner = banner + lines
         if (pylab_o and sympy_o):
             lines = """
-Warning: pylab (numpy and matplotlib) and symbolic math (sympy) are both 
-enabled at the same time. Some pylab functions are going to be overrided by
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/plugins/ipythonconsole/widgets/shell.py]: The script cannot currently ignore individual lines, and doing so would be extremely fragile. Furthermore, if we leave significant trailing spaces, our pre-commit hooks, Spyder's automatic trailing space removal, etc. will remove them if this file is ever touched, causing an undesired functional change (which is the very point of doing all this to begin with).

Therefore, I will instead modify these two lines to return the same output without trailing spaces.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Thanks for your review; blacklisting those files and re-running now. Can ``create_app.py`` be moved to ``tools`` as well, or does it belong in ``scripts`` or the root (as it is now)?

File: spyder/plugins/ipythonconsole/widgets/shell.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/plugins/ipythonconsole/widgets/shell.py]: Actually, is there a reason this whitespace shouldn't be removed? It makes no difference to the output as they are just trailing space before line breaks (that serve no purpose), just like in the source file itself:

With:

image

Without:

image

**Comment by CAM-Gerlach** [comment]
Temporarily closing this so the script doesn't pick it up.

File: .github/ISSUE_TEMPLATE.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply .github/ISSUE_TEMPLATE.md]: Ok, that's even better.

File: spyder/defaults/defaults-2.4.0.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/defaults/defaults-2.4.0.ini]: > Why not?

I meant ini files for Spyder preferences.

File: spyder/plugins/ipythonconsole/widgets/shell.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/widgets/shell.py]: I thought it was necessary for the text to render well. If not, then it's fine to remove it.

**Comment by ccordoba12** [comment]
> Can create_app.py be moved to tools as well

Yes, you can move it there.

> Also, any plans to merge #9226 soon?

I don't understand what's the rush. Can't your script be run again to get rid of the blanks in the files touched by that PR?

----

Also please take of the errors reported by pep8speaks that involve blanks.

**Comment by ccordoba12** [comment]
@CAM-Gerlach, what happened here?

File: spyder/defaults/defaults-2.4.0.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply spyder/defaults/defaults-2.4.0.ini]: I see, I searched the repo and they are all in this dir, so the above regex should take care of them.

**Comment by CAM-Gerlach** [comment]
>  what happened here?

Same as when I closed it before,

> Temporarily closing this so the script doesn't pick it up.

Otherwise, the script would pick up all the files in this PR too, and it was faster and cleaner to close it then add a bespoke exception in the script itself. I didn't repeat that comment again this time since I'd already made it once when I did the same thing previously, which I (mistakenly) thought you would have seen.

> Yes, you can move it there.

Sure, done.

> I don't understand what's the rush.

There's no rush. I just figured it would be easier on both of us to get as many files as possible done in one PR, so we wouldn't have to deal with too many incremental followup PRs, but its fine.

> Also please take care of the errors reported by pep8speaks that involve blanks.

Do you mean all the errors with the word "blank" in them, all of them period, or something more specific? I implemented the first here but if you meant something else, let me know and I'll do that instead.

**Comment by goanpeca** [comment]
probably

>Temporarily closing this so the script doesn't pick it up.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 Re-ran with files/paths added to blacklist, refactored blacklist to make adding items easier, added more robust and friendlier error handling at several points, automatically installed ``pr`` and ``pr-clean`` git aliases if they aren't present, improved documentation, removed blank line PEP 8 issues, general cleanup. Ready for final review.

**Comment by ccordoba12** [comment]
> I just figured it would be easier on both of us to get as many files as possible done in one PR

Ok, I understand. But I'm afraid this is going to require several PRs.

> Do you mean all the errors with the word "blank" in them, all of them period, or something more specific?

Yes, the errors with the word `blank` and the issues reported in `spyder/utils/tests/test_sourcecode.py`

**Comment by CAM-Gerlach** [comment]
> But I'm afraid this is going to require several PRs.

Yeah, I knew it would require more than one, just wanted to minimize the number necessary. But its fine to merge it now.

> Yes, the errors with the word blank and the issues reported in spyder/utils/tests/test_sourcecode.py

The issues with the latter (missing a space after the commas) are no more germane to this PR than the other remaining ones (missing a space after ``#``, too much leading whitespace), so I just fixed all of them except for the line length (which would require more complex/mulitline changes).

File: CHANGELOG.md [hunk 7]
```diff
@@ -446,7 +446,7 @@ In this release 246 pull requests were closed.
 * [Issue 5763](https://github.com/spyder-ide/spyder/issues/5763) - Make arrow-key selection in multiple-option tab-completion dialogs "roll over" ([PR 5771](https://github.com/spyder-ide/spyder/pull/5771))
```
Comment (ccordoba12) [inline CHANGELOG.md]: Please revert the changes to this file and put it in the excluded list. The blanks here come from blanks left in the title of PRs and issues, so I don't want to keep fixing them just for the sake of not avoiding blanks.

Comment (CAM-Gerlach) [reply CHANGELOG.md]: I'm not sure I understand the point, when it takes zero extra effort to have the script remove them along with the others, they still shouldn't be there, they take up extra space in the file and our future commit hook will remove future spaces automatically (or we could just make sure loghub ``.strip()``s issue/PR titles).

Relative to that zero cost and zero downsides, it would require re-writing multiple commits to add it to the blacklist, test that it works, and revert adding the file, for no apparent gain. I'd rather spend that time adding ``.strip()`` to loghub and fixing the problem properly.

Comment (ccordoba12) [reply CHANGELOG.md]: I really think this is going too far. I agree with preventing to add blanks to code files, but this is not code. So, as you did with the Github template files, please add this one to the blacklist.

Comment (goanpeca) [reply CHANGELOG.md]: Agreed this is going great so far, but we should really stick to code files only at the moment. Could you open an issue over loghub to fix the spacing issue? Thanks

Comment (CAM-Gerlach) [reply CHANGELOG.md]: I still don't see a logically articulated reason for that distinction; either Markdown counts as code, or it counts as documentation, the latter of which we've removed spaces from for years, and regardless, they do not change the output. Its just extra work for no apparent gain to exclude it, but not enough to make arguing over it further worthwhile, so I made the change you requested.

For these purposes, there's little point fixing the issue with loghub if we don't fix it here, since there will still be spurious whitespace left in the file if we don't clean them up. Why should we care about loghub not stripping trailing spaces if we don't actually care about them in the output?
